src/plugins/intel_cpu/src/transformations/cpu_opset/common/pass/move_readvalue_inputs_to_subgraph.cpp

```diff
@@ -0,0 +1,129 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/cpu_opset/common/op/read_value_with_subgraph.hpp"
+#include "move_readvalue_inputs_to_subgraph.hpp"
+
+#include <unordered_set>
+
+#include "itt.hpp"
+#include "openvino/core/rt_info.hpp"
+#include "openvino/pass/constant_folding.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "ov_ops/rotary_positional_embeddings.hpp"
+#include "transformations/rt_info/disable_fp16_compression.hpp"
+#include "transformations/utils/gen_pattern.hpp"
+#include "transformations/utils/utils.hpp"
+#include "transformations/cpu_opset/common/op/submodel.hpp"
+
+ov::intel_cpu::MoveReadValueInputsToSubgraph::MoveReadValueInputsToSubgraph() {
+    MATCHER_SCOPE(MoveReadValueInputsToSubgraph);
+    using namespace ov::pass::pattern;
+
+    auto readvalue_pattern = pass::pattern::wrap_type<ov::op::v6::ReadValue>();
+
+    ov::matcher_pass_callback callback = [=](Matcher& m) {
+        const auto& pattern_map = m.get_pattern_value_map();
+        auto readvalue = as_type_ptr<ov::opset6::ReadValue>(pattern_map.at(readvalue_pattern).get_node_shared_ptr());
+        if (!readvalue || readvalue->get_input_size() != 1u) {
+            return false;
+        }
+
+        bool found_output = false;  // Flag: find Output node or out of max depth.
+        std::string root_name = readvalue->get_friendly_name();
+
+        NodeVector subgraph_nodes;
+        std::unordered_set<std::string> subgraph_node_names;
+        std::unordered_set<std::shared_ptr<ov::Node>> visited;  // Visited flag.
+        NodeVector inputs = {};
+        OutputVector outputs = {};
+
+        // DFS, Check if current node's final successor is only ReadValue.
+        std::function<void(std::shared_ptr<ov::Node>)> dfs = [&](std::shared_ptr<ov::Node> node) {
+            if (visited.find(node) != visited.end()) {
+                return;
+            }
+
+            // Output node.
+            if (node->get_output_target_inputs(0).size() == 0u) {
+                found_output = true;
+                return;
+            }
```

Comment:
It does make sense to store the whole DFS subtree, since this is a path to the output node, so checking if a child node is a part of such a subtree may speed up subsequent dfs calls.
In the current implementation we store only nodes visited in reverse_dfs, in the worst case, when we have 1/2N nodes of the ReadValue path all of them parents of the same subgraph leading to an output, so we will have 1/2N * 1/2N iterations, which is still O(N^2). But, if we store the output path as well, all the nodes will be visited once, giving O(N) complexity.

Reply:
Yes. all visited nodes also should be flagged.